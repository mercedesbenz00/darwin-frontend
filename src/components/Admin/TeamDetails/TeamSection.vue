<template>
  <div
    class="section"
    :class="{'section--inactive': !active}"
  >
    <div class="section__title">
      <slot name="title" />
    </div>
    <div class="section__body">
      <slot name="body" />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

@Component({
  name: 'team-section'
})
export default class TeamSection extends Vue {
  @Prop({ required: false, default: true, type: Boolean as () => boolean })
  active!: boolean
}
</script>

<style lang="scss" scoped>
.section {
  @include row;
  flex-wrap: wrap;
  margin-bottom: 25px;
  padding: 20px;
  background: $colorSecondaryLight3;
  border-radius: 3px;

}

.section .section--inactive {
  opacity: .3;
}

.section .section__title {
  width: 80%;
  margin-right: 20%;
  @include typography(xl, headlines, bold);
  margin-bottom: 10px;
}

.section .section__body {
  @include row;
  flex-wrap: wrap;

  gap: 15px;
}
</style>
