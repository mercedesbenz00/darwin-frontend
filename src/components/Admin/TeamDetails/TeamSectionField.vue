<template>
  <div
    class="field"
    :class="{ 'field--full': fullWidth }"
  >
    <label class="field__label">
      <slot name="label" />
    </label>
    <div class="field__value">
      <slot name="value" />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

@Component({
  name: 'team-section-field'
})
export default class TeamSectionField extends Vue {
  @Prop({ required: false, default: false, type: Boolean as () => boolean })
  fullWidth!: boolean
}
</script>

<style lang="scss" scoped>
.field {
  width: 300px;
  @include col;
}

.field.field--full {
  width: 100%;
}

.field.field__label {
  font-weight: bold;
  margin-bottom: 10px;
}
</style>
