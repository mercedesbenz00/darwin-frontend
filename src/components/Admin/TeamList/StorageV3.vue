<template>
  <div class="storage">
    <v-popover trigger="hover">
      <div
        class="storage__info"
        :class="{'storage__info--critical': lowStorage}"
      >
        {{ currentPeriodRemainingGBFormatted }}
      </div>
      <template slot="popover">
        <div class="storage__tooltip-row">
          <label>Billed this period</label>
          <span>{{ currentPeriodBilledGBFormatted }}</span>
        </div>
        <div class="storage__tooltip-row">
          <label>Used this period</label>
          <span>{{ currentPeriodUsedGBFormatted }}</span>
        </div>
        <div class="storage__tooltip-row">
          <label>Billed next period</label>
          <span>{{ nextPeriodBilledGBFormatted }}</span>
        </div>
      </template>
    </v-popover>
  </div>
</template>
<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'

import { CirclePlusIcon } from '@/assets/icons/V1'
import { TeamPayload } from '@/store/modules/admin/types'
import { CustomerSubscriptionPayload } from '@/store/modules/billing/types'
import {
  adaptAdminTeamPayload,
  storageBilledNextPeriod,
  storageBilledThisPeriod,
  storageRemainingThisPeriod
} from '@/utils/billing'

@Component({ name: 'storage', components: { CirclePlusIcon } })
export default class Storage extends Vue {
  @Prop({ required: true })
  team!: TeamPayload

  get subscription (): CustomerSubscriptionPayload {
    if (!this.team.customer_v3) { throw new Error('BillingV3 team missing customer information') }
    return this.team.customer_v3.customer_subscription
  }

  get adaptedInfo (): ReturnType<typeof adaptAdminTeamPayload> {
    return adaptAdminTeamPayload(this.team)
  }

  get currentPeriodBilledGB (): number {
    return storageBilledThisPeriod(this.adaptedInfo)
  }

  get currentPeriodBilledGBFormatted () {
    return this.currentPeriodBilledGB.toLocaleString()
  }

  get currentPeriodUsedGB (): number {
    return this.subscription.storage_used
  }

  get currentPeriodUsedGBFormatted () {
    return this.currentPeriodUsedGB.toLocaleString()
  }

  get currentPeriodRemainingGB (): number {
    return storageRemainingThisPeriod(this.adaptedInfo)
  }

  get currentPeriodRemainingGBFormatted () {
    return this.currentPeriodRemainingGB.toLocaleString()
  }

  get lowStorage (): boolean {
    return this.currentPeriodRemainingGB < 5
  }

  get nextPeriodBilledGB (): number {
    return storageBilledNextPeriod(this.adaptedInfo)
  }

  get nextPeriodBilledGBFormatted () {
    return this.nextPeriodBilledGB.toLocaleString()
  }
}
</script>

<style lang="scss" scoped>
.storage {
  @include row--right;
  width: 100%;
}

.storage__info {
  margin-right: 10px;
}

.storage-storage__info--critical {
  font-weight: bold;
  color: $colorPink;
}

.storage__tooltip-row {
  min-width: 150px;
  margin: 10px;
  @include row--distributed;
}

.storage__tooltip-row label {
  font-weight: bold;
}
</style>
