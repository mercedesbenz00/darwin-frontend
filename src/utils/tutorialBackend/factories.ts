import {
  AnnotateStagePayload,
  CodeStagePayload,
  CompleteStagePayload,
  DatasetImagePayload,
  DatasetItemPayload,
  DatasetItemStatus,
  DatasetItemType,
  ImagePayload,
  ModelStagePayload,
  ReviewStagePayload,
  StageType,
  WorkflowStagePayload,
  WorkflowStageTemplatePayload,
  WorkflowTemplatePayload
} from '@/store/types'

import TutorialError from './TutorialError'
import { dataset } from './data/dataset'

const itemId = (datasetImageId: number) => datasetImageId * 10
const workflowId = (datasetImageId: number) => datasetImageId * 100

const stageId = (datasetImageId: number, stageNumber: number) =>
  workflowId(datasetImageId) + stageNumber

const initializeDatasetImage = (image: ImagePayload): DatasetImagePayload => ({
  dataset_id: dataset.id,
  dataset_video_id: null,
  id: image.id * 10,
  image,
  seq: image.id,
  set: 1
})

const initializeItem = (datasetImage: DatasetImagePayload): DatasetItemPayload => ({
  archived_reason: null,
  archived: false,
  current_workflow: null,
  current_workflow_id: null,
  dataset_id: datasetImage.dataset_id,
  dataset_image_id: datasetImage.id,
  dataset_image: datasetImage,
  dataset_video_id: null,
  dataset_video: null,
  file_size: 0,
  filename: datasetImage.image.original_filename,
  height: datasetImage.image.height,
  id: itemId(datasetImage.id),
  inserted_at: '2020-06-12T00:00:00',
  labels: [],
  path: '/',
  priority: 0,
  seq: datasetImage.seq,
  set: datasetImage.set,
  status: DatasetItemStatus.annotate,
  updated_at: '2020-06-12T00:00:00',
  width: datasetImage.image.width,
  type: DatasetItemType.image
})

const initializeStage = (
  datasetImage: DatasetImagePayload,
  stageTemplate: WorkflowStageTemplatePayload
): WorkflowStagePayload => {
  const { type } = stageTemplate
  const payload = {
    assignee_id: null,
    completed: false,
    completes_at: null,
    dataset_item_id: itemId(datasetImage.id),
    id: stageId(datasetImage.id, stageTemplate.stage_number),
    metadata: {},
    number: stageTemplate.stage_number,
    skipped_reason: null,
    skipped: false,
    template_metadata: stageTemplate.metadata,
    type: stageTemplate.type,
    workflow_id: workflowId(datasetImage.id),
    workflow_stage_template_id: stageTemplate.id
  }

  if (type === StageType.Annotate) { return payload as AnnotateStagePayload }
  if (type === StageType.Code) { return payload as CodeStagePayload }
  if (type === StageType.Complete) { return payload as CompleteStagePayload }
  if (type === StageType.Model) { return payload as ModelStagePayload }
  if (type === StageType.Review) { return payload as ReviewStagePayload }

  throw new Error('Invalid type given')
}

const initializeWorkflow = (
  datasetImage: DatasetImagePayload,
  workflowTemplate: WorkflowTemplatePayload,
  currentStage: number
): Pick<DatasetItemPayload, 'current_workflow' | 'current_workflow_id'> => {
  const annotateTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Annotate)

  const reviewTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Review)

  const completeTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Complete)

  if (!annotateTemplate || !reviewTemplate || !completeTemplate) {
    throw new TutorialError('Initializing workflow using invalid template ')
  }

  const id = workflowId(datasetImage.id)

  const workflow: DatasetItemPayload['current_workflow'] = {
    dataset_item_id: itemId(datasetImage.id),
    id,
    workflow_template_id: workflowTemplate.id,
    current_stage_number: currentStage,
    current_workflow_stage_template_id: annotateTemplate.id,
    status: DatasetItemStatus.annotate,
    stages: {
      1: [initializeStage(datasetImage, annotateTemplate)],
      2: [initializeStage(datasetImage, reviewTemplate)],
      3: [initializeStage(datasetImage, completeTemplate)]
    }
  }

  return { current_workflow: workflow, current_workflow_id: id }
}

export const createItem = (
  image: ImagePayload,
  workflowTeplate: WorkflowTemplatePayload,
  stageNumber: number = 1
): DatasetItemPayload => {
  const datasetImage = initializeDatasetImage(image)
  return {
    ...initializeItem(datasetImage),
    ...initializeWorkflow(datasetImage, workflowTeplate, stageNumber)
  }
}
