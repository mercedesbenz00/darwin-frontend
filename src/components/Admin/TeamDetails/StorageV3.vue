<template>
  <team-section
    class="storage-v3"
    :active="active"
  >
    <template #title>
      Storage (V3)
    </template>
    <template #body>
      <team-section-field>
        <template #label>
          Billed this period
        </template>
        <template #value>
          {{ format(billedThisPeriod) }}
        </template>
      </team-section-field>
      <team-section-field>
        <template #label>
          Billed next period
        </template>
        <template #value>
          {{ format(billedNextPeriod) }}
        </template>
      </team-section-field>
      <team-section-field>
        <template #label>
          Extra storage
        </template>
        <template #value>
          <extra-storage :team="team" />
        </template>
      </team-section-field>
      <team-section-field>
        <template #label>
          Used
        </template>
        <template #value>
          {{ format(used) }}
        </template>
      </team-section-field>
      <team-section-field>
        <template #label>
          Remaining this period
        </template>
        <template #value>
          {{ format(remainingThisPeriod) }}
        </template>
      </team-section-field>
    </template>
  </team-section>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'

import { TeamPayload } from '@/store/modules/admin/types'
import { CustomerSubscriptionPayload } from '@/store/modules/billing/types'
import {
  storageBilledThisPeriod,
  adaptAdminTeamPayload,
  storageBilledNextPeriod,
  storageRemainingThisPeriod
} from '@/utils/billing'

import ExtraStorage from './ExtraStorage.vue'
import TeamSection from './TeamSection.vue'
import TeamSectionField from './TeamSectionField.vue'

@Component({
  name: 'storage-v3',
  components: { ExtraStorage, TeamSection, TeamSectionField }
})
export default class CreditsV extends Vue {
  @Prop({ required: true, type: Object as () => TeamPayload })
  team!: TeamPayload

  get active (): boolean {
    return !!this.team.customer_v3
  }

  get subscription (): CustomerSubscriptionPayload | null {
    return this.team.customer_v3 ? this.team.customer_v3.customer_subscription : null
  }

  get adaptedInfo (): ReturnType<typeof adaptAdminTeamPayload> {
    return adaptAdminTeamPayload(this.team)
  }

  get billedThisPeriod (): number {
    return storageBilledThisPeriod(this.adaptedInfo)
  }

  get billedNextPeriod (): number {
    return storageBilledNextPeriod(this.adaptedInfo)
  }

  get used (): number {
    const { subscription } = this
    return subscription ? subscription.storage_used : 0
  }

  get remainingThisPeriod (): number {
    return storageRemainingThisPeriod(this.adaptedInfo)
  }

  format (value: number): string {
    return value.toLocaleString(undefined, { maximumFractionDigits: 3 })
  }
}
</script>
