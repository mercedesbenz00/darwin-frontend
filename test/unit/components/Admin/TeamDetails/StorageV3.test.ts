import { shallowMount, createLocalVue } from '@vue/test-utils'

import { buildCustomerSubscriptionPayload } from 'test/unit/factories'
import { buildAdminTeamPayload } from 'test/unit/factories/buildAdminTeamPayload'

import StorageV3 from '@/components/Admin/TeamDetails/StorageV3.vue'
import TeamSection from '@/components/Admin/TeamDetails/TeamSection.vue'
import TeamSectionField from '@/components/Admin/TeamDetails/TeamSectionField.vue'

const localVue = createLocalVue()
const periodStart = 1000

const v7 = buildAdminTeamPayload({
  name: 'V7',
  subscription_period_start: periodStart,
  customer_v3: {
    customer_subscription: buildCustomerSubscriptionPayload({
      storage_standard: 100,
      storage_standard_max_in_period: 200,
      storage_standard_max_in_period_at: periodStart,
      storage_used: 50
    })
  }
})

it('matches snapshot', () => {
  const propsData = { team: v7 }
  const stubs = { TeamSection, TeamSectionField }
  const wrapper = shallowMount(StorageV3, { localVue, propsData, stubs })
  expect(wrapper).toMatchSnapshot()
})
