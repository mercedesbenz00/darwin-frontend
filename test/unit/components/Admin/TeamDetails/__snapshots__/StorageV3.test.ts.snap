// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div class="section storage-v3">
  <div class="section__title">
    Storage (V3)
  </div>
  <div class="section__body">
    <div class="field"><label class="field__label">
        Billed this period
      </label>
      <div class="field__value">
        200
      </div>
    </div>
    <div class="field"><label class="field__label">
        Billed next period
      </label>
      <div class="field__value">
        100
      </div>
    </div>
    <div class="field"><label class="field__label">
        Extra storage
      </label>
      <div class="field__value">
        <extra-storage-stub team="[object Object]"></extra-storage-stub>
      </div>
    </div>
    <div class="field"><label class="field__label">
        Used
      </label>
      <div class="field__value">
        50
      </div>
    </div>
    <div class="field"><label class="field__label">
        Remaining this period
      </label>
      <div class="field__value">
        150
      </div>
    </div>
  </div>
</div>
`;
